dart-mode.el

```diff
@@ -562,6 +579,251 @@ This uses `dart-format-path' to find the formatter."
+  "Construct a request for the analysis server.
+
+The constructed request will call METHOD with optional PARAMS."
+  `((method . ,method) (params . ,params)))
+
+(defun dart-analysis-server-enqueue (req-without-id callback)
+  "Send REQ-WITHOUT-ID to the analysis server, call CALLBACK with the result."
+  (setq dart-analysis-server-next-id (1+ dart-analysis-server-next-id))
+  (let ((request
+         (json-encode (push `(id . ,(format "%s" dart-analysis-server-next-id))
+                            req-without-id))))
+    (dart-log (concat "Sent:\n" request))
+    (if callback
+        (push `(,dart-analysis-server-next-id . ,callback)
+              dart-analysis-server-callbacks)
+      (push
+       `(,dart-analysis-server-next-id . dart-analysis-server-no-error-callback)
+       dart-analysis-server-callbacks))
+    (process-send-string (dart-analysis-server-process dart-analysis-server)
+                         (concat request "\n"))))
+
+(defun dart-analysis-server-no-error-callback (response)
+  "If RESPONSE has an error, report it."
+  (-when-let (resp-err (assoc 'error response))
+    (dart-log (format "Response from server had error: %s" resp-err))))
+
+(defun dart-analysis-server-handle-response (callback response)
+  "Call CALLBACK with the parsed JSON RESPONSE from the analysis server."
+  (dart-log (concat "Received:\n" (format "%s" response)))
+  (funcall callback response))
+
+(defun dart-analysis-server-process-filter (das string)
+  "Handle the event or method response from the dart analysis server.
+
+The server DAS has STRING added to the buffer associated with it.
+Method responses are paired according to their pending request and
+the callback for that request is given the json decoded response."
+  (let ((buf (dart-analysis-server-buffer das)))
+    (when (buffer-live-p buf)
+      (with-current-buffer buf
+        (goto-char (point-max))
+        (insert string)
+        (let ((buf-lines (split-string (buffer-string) "\n")))
+          (delete-region (point-min) (point-max))
+          (insert (-last-item buf-lines))
+
+          (let ((json-lines (-map 'json-read-from-string
+                                  (-filter 's-present? (-butlast buf-lines)))))
+            (-each json-lines 'dart-analysis-server-handle-msg)))))))
+
+(defun dart-analysis-server-handle-msg (msg)
+  "Handle the parsed MSG from the analysis server."
+  (-when-let* ((id-assoc (assoc 'id msg))
+               (raw-id (cdr id-assoc))
+               (id (string-to-number raw-id)))
+    (-if-let (resp-closure (assoc id dart-analysis-server-callbacks))
+        (progn
+          (dart-analysis-server-handle-response (cdr resp-closure) msg)
+          (setq dart-analysis-server-callbacks
+                (assq-delete-all id dart-analysis-server-callbacks)))
+      (dart-log (format "No callback was associated with id %s" raw-id)))))
+
+(cl-defstruct dart-analysis-server process buffer queue)
```

Comment:
Move this definition to the top of the "Analysis Server" section.

Users shouldn't interact with this manually, so you should make the constructor and accessors private:

```elisp
(cl-defstruct (dart-analysis-server
               (:conc-name dart--analysis-server)
               (:constructor dart--make-analysis-server))
  process buffer queue)
```

Also, make sure you document the fields.